src/sparse_blas/backends/cusparse/operations/cusparse_spmm.cpp

```diff
@@ -0,0 +1,296 @@
+                        oneapi::mkl::sparse::dense_matrix_handle_t B_handle, const void* beta,
+                        oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+                        oneapi::mkl::sparse::spmm_alg alg, void* workspace_ptr,
+                        bool is_alpha_host_accessible) {
+    auto cu_a = A_handle->backend_handle;
+    auto cu_b = B_handle->backend_handle;
+    auto cu_c = C_handle->backend_handle;
+    auto type = A_handle->value_container.data_type;
+    auto cu_op_a = get_cuda_operation(type, opA);
+    auto cu_op_b = get_cuda_operation(type, opB);
+    auto cu_type = get_cuda_value_type(type);
+    auto cu_alg = get_cuda_spmm_alg(alg);
+    set_pointer_mode(cu_handle, is_alpha_host_accessible);
+    auto status = cusparseSpMM_preprocess(cu_handle, cu_op_a, cu_op_b, alpha, cu_a, cu_b, beta,
+                                          cu_c, cu_type, cu_alg, workspace_ptr);
+    check_status(status, "optimize_spmm");
+}
+
+void spmm_optimize(sycl::queue& queue, oneapi::mkl::transpose opA, oneapi::mkl::transpose opB,
+                   const void* alpha, oneapi::mkl::sparse::matrix_view A_view,
+                   oneapi::mkl::sparse::matrix_handle_t A_handle,
+                   oneapi::mkl::sparse::dense_matrix_handle_t B_handle, const void* beta,
+                   oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+                   oneapi::mkl::sparse::spmm_alg alg, oneapi::mkl::sparse::spmm_descr_t spmm_descr,
+                   sycl::buffer<std::uint8_t, 1> workspace) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    if (!A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    common_spmm_optimize(opA, opB, is_alpha_host_accessible, A_view, A_handle, B_handle,
+                         is_beta_host_accessible, C_handle, alg, spmm_descr);
+    // Copy the buffer to extend its lifetime until the descriptor is free'd.
+    spmm_descr->workspace.set_buffer_untyped(workspace);
+    if (alg == oneapi::mkl::sparse::spmm_alg::no_optimize_alg || workspace.size() == 0) {
+        // cusparseSpMM_preprocess cannot be called if the workspace is empty
+        return;
+    }
+    fallback_alg_if_needed(alg, opA, opB);
+    auto functor = [=](CusparseScopedContextHandler& sc,
+                       sycl::accessor<std::uint8_t> workspace_acc) {
+        auto cu_handle = sc.get_handle(queue);
+        auto workspace_ptr = sc.get_mem(workspace_acc);
+        spmm_optimize_impl(cu_handle, opA, opB, alpha, A_handle, B_handle, beta, C_handle, alg,
+                           workspace_ptr, is_alpha_host_accessible);
+    };
+
+    sycl::accessor<std::uint8_t, 1> workspace_placeholder_acc(workspace);
+    dispatch_submit(__func__, queue, functor, A_handle, workspace_placeholder_acc, B_handle,
+                    C_handle);
```

Comment:
Is there a reason why a placeholder `sycl::accessor` is created in advance here and bound to the command group `handler` through a `cgh::require()` call later, instead of passing the `sycl::buffer` workspace along inside the dispatch function and directly creating a regular accessor on the workspace inside the `host_task` in `submit_host_task_with_acc()` function? I'm not saying we shouldn't do what you have, but it definitely seems like advanced usage; I and some others I know haven't come across such usage before.

Reply:
The placeholder access can be default constructed with no cost whereas the buffer requires a range. I see the buffer size is allowed to be 0 now, I avoided it out of habit as this was not always supported.
This is useful when calling [`dispatch_submit`](https://github.com/oneapi-src/oneMKL/pull/527/files#diff-4520860f5bfcef6df30df555408192906db7a83056c4fd75018d5f99b7de4e6dR319) when the workspace argument is not needed. This may be rewritten without the placeholder accessor but I don't think the code will look any simpler to be honest.

Reply:
If you delay the accessor creation until the `host_task` submission, that would just mean you pass in the `sycl::buffer` as-is deeper and deeper and finally create an accessor with `buf->template get_access()` in the `queue.submit()` call instead of passing in `sycl::accessor` deeper and then binding it to the command group handler, isn't it?

_If_ that's the only difference and there are no other code complications, I would lean toward making the change to passing in the `sycl::buffer` deeper as it is beginner-friendly.

Reply:
That was easier than expected, it's done in https://github.com/oneapi-src/oneMKL/pull/527/commits/4db2187e21ca99000cbab08cafa768e94e5df14a

Reply:
Great, that was what I was hoping for! Thanks for changing this.